### PR TITLE
[TTI] NFC: Port TLI.shouldSinkOperands to TTI

### PR DESCRIPTION
**Pull request by jrbyrnes**
Porting to TTI provides direct access to the instruction cost model, which can enable instruction cost based sinking without introducing code duplication.

Added some folks from various backends this touches for any backend specific concerns.

### CODE REVIEW

File: llvm/include/llvm/Analysis/TargetTransformInfo.h [hunk 1]
```diff
@@ -1744,6 +1744,10 @@ class TargetTransformInfo {
   bool hasActiveVectorLength(unsigned Opcode, Type *DataType,
                              Align Alignment) const;
 
+  /// \return True if the operands of \p I should be sunk to \p I . The sinkable
+  /// operands are cached in \p OpsToSink
```
Comment (arsenm) [inline llvm/include/llvm/Analysis/TargetTransformInfo.h:1748]: This lost some of the original comment

File: llvm/include/llvm/Analysis/TargetTransformInfoImpl.h [hunk 2]
```diff
@@ -968,6 +968,10 @@ class TargetTransformInfoImplBase {
     return false;
   }
 
+  bool shouldSinkOperands(Instruction *I, SmallVectorImpl<Use *> &Ops) const {
+    return false;
```
Comment (arsenm) [inline llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:972]: Maybe checking for free ops should be the BasicTTI default?

File: llvm/include/llvm/Analysis/TargetTransformInfo.h [hunk 3]
```diff
@@ -1744,6 +1744,13 @@ class TargetTransformInfo {
   bool hasActiveVectorLength(unsigned Opcode, Type *DataType,
                              Align Alignment) const;
 
+  /// Return true if sinking I's operands to the same basic block as I is
+  /// profitable, e.g. because the operands can be folded into a target
+  /// instruction during instruction selection. After calling the function
+  /// \p Ops contains the Uses to sink ordered by dominance (dominating users
+  /// come first).
+  bool shouldSinkOperands(Instruction *I, SmallVectorImpl<Use *> &Ops) const;
```
Comment (arsenm) [inline llvm/include/llvm/Analysis/TargetTransformInfo.h:1752]: I think the name needs work if it's moving to a more general place

File: llvm/lib/Target/X86/X86TargetTransformInfo.cpp [hunk 4]
```diff
@@ -6894,3 +6894,82 @@ InstructionCost X86TTIImpl::getBranchMispredictPenalty() const {
   // TODO: Hook MispredictPenalty of SchedMachineModel into this.
   return 14;
 }
+
+static bool isVectorShiftByScalarCheap(Type *Ty, const X86Subtarget *ST) {
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86TargetTransformInfo.cpp:6898]: Maybe move TargetLowering::isVectorShiftByScalarCheap into TTI as well? X86 is its only user

Comment (jrbyrnes) [reply llvm/lib/Target/X86/X86TargetTransformInfo.cpp]: Sure, I've moved it, though I think the function needs a minor rework if we wanted to use the cost model as its implementation.

**Review by arsenm — APPROVED** [review]

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Pull request merged**